### PR TITLE
[FLINK-34537] Autoscaler JDBC Support HikariPool

### PR DESCRIPTION
**Pull request by czy006**
## What is the purpose of the change

*Autoscaler JDBC Support HikariPool to Replace JDBC DirverManager*

## Brief change log

  - *Add Autoscaler HikariPool Test*
  - *Replace JDBC DirverManager Wiht HikariPool DataSource*

## Verifying this change

  - *Added HikariPool Connection Test*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: (no)
  - Core observer or reconciler logic that is regularly executed: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by czy006** [comment]
> Thanks @czy006 for the improvement!
> 
> I'm on vacation recently, I will review it after it. Sorry for the delay.
> 
> Btw, `1.8.0` will be released soon, would you mind merging it in the next version?

no problem

File: flink-autoscaler-plugin-jdbc/pom.xml
Comment (1996fanrui) [inline flink-autoscaler-plugin-jdbc/pom.xml:147]: 1. The `HikariCP` is production dependency, it's better to move it before test related dependencies.
2. The version is better to extract a property.

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (1996fanrui) [inline flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:40]: The user and password can be null, did you test it with Autoscaler Standalone?

Also, could AutoscalerStateStoreFactoryTest use this method as well instead of create `HikariConfig` and `HikariDataSource` directly?

**Review by 1996fanrui — COMMENTED** [review]
Thanks @czy006 for the improvement!

Overall is good to me, could you test it with autoscaler standalone process in your local?

Also, I see the CI fails, please help fix it as well, thanks a lot~

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (1996fanrui) [inline flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:32]: I saw all callers of `getHikariDataSource` calls the `getConnection()` directly.

Could we return Connection directly? It means we rename this method to `getConnection`.

Please correct me if I missed anything, thanks~

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (czy006) [reply flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:43]: Now that I've added it to my tests, Derby works fine with its empty user name and password, and I didn't test the other databases

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (czy006) [reply flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java]: now include this part of the test in unit tests

File: flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerStateStoreFactoryTest.java
Comment (1996fanrui) [inline flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerStateStoreFactoryTest.java:66]: This test class is related to state store instead of event handler.

Please check the comment, `STATE_STORE_TYPE` and `testCreateJdbcEventHandler`.

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (1996fanrui) [inline flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:33]: This class doesn't have any comment. It will break the CI.

**Review by 1996fanrui — COMMENTED** [review]
Hi @czy006  thanks for the update!

I left 2 minor comments, please take a look in your free time. 

Also, please ensure the CI is green after your change, thanks a lot.

File: flink-autoscaler-standalone/pom.xml
Comment (1996fanrui) [inline flink-autoscaler-standalone/pom.xml:166]: It's still after the Test related dependencies. You can check `<!-- Test -->` comment, you can move `HikariCP` before it.

**Review by 1996fanrui — COMMENTED** [review]
The CI is still failed, the root cause is `HikariCP-java7-2.4.13.jar, HikariCP-5.1.0.jar define 59 overlapping classes: `.

autoscaler depends on `HikariCP-java7` before, I'm not sure whether we need the `HikariCP` in this PR?

Also, would you mind testing autoscaler standalone manually before merging to ensure it works well with MySQL and PostgreSQL?  I will test manually as well these 2 days.

File: flink-autoscaler/pom.xml
Comment (1996fanrui) [inline flink-autoscaler/pom.xml:71]: Are you sure this dependency works with quartz? Did you test flink-kubernetes-operator with autoscaler?

File: flink-autoscaler-standalone/pom.xml
Comment (1996fanrui) [inline flink-autoscaler-standalone/pom.xml:140]: If `flink-autoscaler` depends on `HikariCP`, and `flink-autoscaler-standalone` depends on `flink-autoscaler`.

Why does `flink-autoscaler-standalone` still need to depends on `HikariCP`? IIUC,  `flink-autoscaler-standalone`  already depends on `HikariCP` silently.

Please correct me if my understanding is wrong.

File: flink-autoscaler/pom.xml
Comment (czy006) [reply flink-autoscaler/pom.xml:71]: Thank you for your feedback. I'll be testing MySQL and PostgreSQL later, and I'm still tweaking it.

File: flink-autoscaler-standalone/pom.xml
Comment (czy006) [reply flink-autoscaler-standalone/pom.xml:140]: I first passed the compilation, ran it locally myself and again saw if it worked as expected. I'm trying to remove that dependency to see if it's affected, and I'll come to a conclusion later

File: flink-autoscaler/pom.xml
Comment (1996fanrui) [reply flink-autoscaler/pom.xml:71]: >  I'll be testing MySQL and PostgreSQL later, and I'm still tweaking it.

So far, flink-kubernetes-operator doesn't support Jdbc state store and event handler. But autoscaler uses quartz, this comment isn't related to MySQL and PostgreSQL.

This comment means whether flink-kubernetes-operator with autoscaler works well when we introduce the other version for `HikariCP`.

Also, standalone-autoscaler will use the Jdbc state store and event handler, it's needed to test them.

File: flink-autoscaler-standalone/pom.xml
Comment (1996fanrui) [reply flink-autoscaler-standalone/pom.xml:140]: Hi @czy006 , this PR is not updated more than 6 weeks, is there any conclusion about this comment and this PR?

**Review by 1996fanrui — APPROVED** [review]
I updated some old comments due to @czy006 doesn't response for a long time.

I ran it locally, the autoscaler standalone works well.

Hi @RocMarshal , could you help review this PR as well? thanks in advance.

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (RocMarshal) [inline flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:39]:
```suggestion
               !StringUtils.isNullOrWhitespaceOnly(jdbcUrl),
```

File: flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java
Comment (RocMarshal) [inline flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/utils/HikariJDBCUtil.java:40]: what about extracting it as a static final constant for fine-test & reference ?
eg:
```
public static final String JDBC_URL_REQUIRED_HINT_FORMAt = "%s is required when jdbc state store or jdbc event handler is used.";
```
If so, we could use the constant in the corresponding test cases. `anchor A1 & anchor A2`

File: flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerEventHandlerFactoryTest.java
Comment (RocMarshal) [inline flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerEventHandlerFactoryTest.java:61]: anchor A1

File: flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerStateStoreFactoryTest.java
Comment (RocMarshal) [inline flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/AutoscalerStateStoreFactoryTest.java:60]: anchor A2

**Review by RocMarshal — COMMENTED** [review]
Thanks @czy006  for the contribution and @1996fanrui for the review.
LGTM on the whole.
Just left a few of minor comments.
PATL if you had the free time. :)

**Comment by 1996fanrui** [comment]
> Thanks @czy006 for the contribution and @1996fanrui for the review. LGTM on the whole. Just left a few of minor comments. PATL if you had the free time. :)

Thanks @RocMarshal for the quick review, I have updated all your comments.

**Review by RocMarshal — APPROVED** [review]
Thanks @1996fanrui for the update.
LGTM +1 assuming the CI is successful.

**Pull request merged**
